Create 1-introduction.md (#31)
Vaibhavpratapsingh22/Courses: 4-software-engineering/1-introduction.md

#  Definition of Software

Software means a collection of programs whose objective is to enhance the capabilities of the hardware. So software can be defned  as :

**Software is a collection of computer programs , procedure rules and associated documentation and data.**

## Program Versus Software
It is clear software is not merely collection of computer programs. Any program is a subset of software and it becomes a software ony if documentation and operating procedure manuals are prepared.

There are three mai components of software :

1. Program 
2. Documentaion
3. Operating Procedures

Hence ,

> **Software= Program + Documentaion + Operating Procedures**

Reviewer: Isnt operating procedures included in documentation already?

Author: No , Documentation is that which contains different types of manuals like flow-charts , entity relation diagrams whereas operating procedure consists of instructions to setup and how to use the software and how to react to system failure. For eg -installation guides ,  reference guide etc .

Reviewer: like a User Manual you mean?

Author: Yes

The quality of software was low quality
Software often did not meet requirements
The average software project overshoots its schedule by half
At that time Software was never delivered
Solution of Software Crisis:
There is no single solution to the crisis.one possible solution of software crisis is Software Engineering because software engineering is a systematic, disciplined and quantifiable approach. For preventing software crisis, there are some guidelines:

1. Reduction in software over-budget
2. The quality of software must be high
3. Less time needed for software project
4. Experience working team member on software project
5. Software must be delivered on time

# Software Characterstics
Software Characteristics are classified into six major components:

![components](https://media.geeksforgeeks.org/wp-content/uploads/20190321145114/Untitled-Diagram-91.png)

These components are described below:

### 1. Functionality:
It refers to the degree of performance of the software against its intended purpose.
Required functions are:
![functionality](https://media.geeksforgeeks.org/wp-content/uploads/20190321135958/Untitled-Diagram31.png)

### 2. Reliability:

> It refers to the ability of software to work which it is supposed to do . The software should continue to operate without interruption when one or more of its components fail. Software should become more reliable w.r.t time .

Reviewer: ```suggestion
It refers to the ability of software to work which it is supposed to do. The software should continue to operate without interruption when one or more of its components fail. The software should become more reliable with respect to time.
```

> ![reliability](https://media.geeksforgeeks.org/wp-content/uploads/20190321141051/Untitled-Diagram-42.png)

Reviewer: The text is overflowing image here.

### 3. Efficiency:
It refers to the ability of the software to use system resources in the most effective and efficient manner.the software should make effective use of storage space and executive command as per desired timing requirement.
![efficiency](https://media.geeksforgeeks.org/wp-content/uploads/20190321141051/Untitled-Diagram-42.png)

### 4. Usability:
It refers to the extent to which the software can be used with ease.the amount of effort or time required to learn how to use the software.

> ![usability](https://media.geeksforgeeks.org/wp-content/uploads/20190321142138/Untitled-Diagram-61.png)

Reviewer: 4,5,6 text is overflowing images. Maybe images could be reduced in size as well. They are looking too big for now.

### 5. Maintainability:
It refers to the ease with which the modifications can be made in a software system to extend its functionality, improve its performance, or correct errors.

> ![maintainability](https://media.geeksforgeeks.org/wp-content/uploads/20190321142611/Untitled-Diagram-71.png)

Reviewer: Nice consistent images.

### 6. Portability:
A set of attribute that bear on the ability of software to be transferred from one environment to another, without or minimum changes.
![portability](https://media.geeksforgeeks.org/wp-content/uploads/20190321143347/Untitled-Diagram-81.png)

> 

Reviewer: Please provide sources of images as well as the content. Lets keep the work genuine and give equal credits to the creators. 
Citation.

# Definition of Software Engineering

Software Engeneering is discipline whose aim is the production of fault free software that satisfy the user's need and that is delivered on time and within the budget.

From the above definition we can say that the **aim or goal** of software engeneering is to produce software within following attributes :

1. High quality
2. Low cost
3. Within time
4. Consistency
5. Efficiency
6. Extensibility 
7. Satisfying the requirements

## Elements of Software Engeneering

Elements of Software Engeneering are :

> 1. Methods : A method is a procedure for producing some results.

Reviewer: I didnt get what are the elements of software engineering. Please try to explain it in your words

> 1. Methods : A method is a procedure for producing some results.

Reviewer: This section is same - are we keeping it like this only?

Author: Let it be same for now , we will update it later on when i will find some better stuff on this .

2. Procedures : A procedure is the combination of tools and menthods to produce a particular product. For example Algorihtm.
3. Tools : Tools are used to increase 
accuracy , efficiency , productivity , quality of the end product.
4. Paradigms : Paradigms reffer to particular approaches or philosipphies for building software . For example object oriented paradigm .

> 

Reviewer: Closing part is missing here as well. The document suddenly ends.

> ![edit2](https://img.shields.io/static/v1?label=Source&message=www.geeksforgeeks.org&color=orange)![edit](https://img.shields.io/static/v1?label=PRs&message=Welcome&color=<COLOR>)[![Awesome](https://cdn.rawgit.com/sindresorhus/awesome/d7305f38d29fed78fa85652e3a63e154dd8e8829/media/badge.svg)](https://github.com/sindresorhus/awesome#readme)

Reviewer: Source is added - cool

![portability](https://media.geeksforgeeks.org/wp-content/uploads/20190321143347/Untitled-Diagram-81.png)

# Definition of Software Engineering

Software Engeneering is discipline whose aim is the production of fault free software that satisfy the user's need and that is delivered on time and within the budget.

From the above definition we can say that the **aim or goal** of software engeneering is to produce software within following attributes :

1. High quality
2. Low cost
3. Within time
4. Consistency
5. Efficiency
6. Extensibility 
7. Satisfying the requirements

## Elements of Software Engeneering

Elements of Software Engeneering are :

1. Methods : A method is a procedure for producing some results.
2. Procedures : A procedure is the combination of tools and menthods to produce a particular product. For example Algorihtm.
3. Tools : Tools are used to increase 
accuracy , efficiency , productivity , quality of the end product.
4. Paradigms : Paradigms reffer to particular approaches or philosipphies for building software . For example object oriented paradigm .

## Summary

> In this chapter we will look at the basics of software engineering .We will start with how a program is different from a software . Then we will head towards the software crisis . In software crisis we will learn its causes , characteristics . At last we will learn about the elments of software engineering .

Reviewer: ```suggestion
In this chapter, we had a look at the basics of software engineering. We started with how a program is different from software. Then we headed towards the software crisis where we learned its causes, characteristics. At last, we quickly went over the elements of software engineering.
```
